[LLVM][TableGen] Code cleanup in FastISelEmitter.cpp

jurahul opened the pull request:
- Use StringRef instead of std::string for `InstructionMemo::Name`.
- Use range for loops, zip_equal and structured bindings in loops.
- Use llvm::any_of instead of manual loops.
- Use ListSeparator.
- Remove {} around single-line if-else chains.
- Use ArrayRef<> instead of const vector reference for function args.
- Change `getLegalCName` to accept a `StringRef` to avoid StringRef->std::string casting in several places.
- Use StringRef instead of std::string for `OpcodeName` (and in associated maps).

Tested by verifying no changes in .inc files with and without this change.

jurahul committed:
[LLVM][TableGen] Code cleanup in FastISelEmitter.cpp

- Use StringRef instead of std::string for `InstructionMemo::Name`.
- Use range for loops, zip_equal and structured bindings in loops.
- Use llvm::any_of instead of manual loops.
- Use ListSeparator.
- Remove {} around single-line if-else chains.
- Use ArrayRef<> instead of const vector reference for function args.
- Change `getLegalCName` to accept a `StringRef` to avoid
  StringRef->std::string casting in several places.
- Use StringRef instead of std::string for `OpcodeName` (and in
  associated maps).

Tested by verifying no changes in .inc files with and without this
change.

wangpc-pp commented on llvm/utils/TableGen/FastISelEmitter.cpp:
```diff
@@ -71,10 +71,7 @@ class ImmPredicateSet {
-    assert(i < PredsByName.size());
-    return PredsByName[i];
-  }
+  const TreePredicateFn &getPredicate(unsigned Idx) { return PredsByName[Idx]; }
```
Why do we remove this assert?

s-barannikov commented on llvm/utils/TableGen/FastISelEmitter.cpp:
```diff
@@ -151,48 +148,42 @@ struct OperandsSignature {
-        Result.Operands.push_back(OpKind::getImm(0));
+    Result.Operands.resize(Operands.size());
+    for (auto [RO, O] : zip_equal(Result.Operands, Operands))
+      RO = O.isImm() ? OpKind::getImm(0) : O;
```
`std::transform` would be idiomatic here.

jurahul replied:
Done.

s-barannikov commented on llvm/utils/TableGen/FastISelEmitter.cpp:
```diff
@@ -304,77 +295,74 @@ struct OperandsSignature {
-    for (unsigned i = 0, e = Operands.size(); i != e; ++i) {
-      if (PR[i] != "")
+    for (const auto [Idx, Opnd, PhyReg] : enumerate(Operands, PhyRegs)) {
+      if (PhyReg != "") {
```
```suggestion
      if (!PhyReg.empty()) {
```

jurahul replied:
Done

s-barannikov commented on llvm/utils/TableGen/FastISelEmitter.cpp:
```diff
@@ -304,77 +295,74 @@ struct OperandsSignature {
-    for (unsigned i = 0, e = Operands.size(); i != e; ++i) {
-      if (PR[i] != "")
+    for (const auto [PR, Opnd] : zip_equal(PhyRegs, Operands)) {
+      if (PR != "") {
```
```suggestion
      if (!PR.empty()) {
```

jurahul replied:
Done

s-barannikov commented on llvm/utils/TableGen/FastISelEmitter.cpp:
```diff
@@ -304,77 +295,74 @@ struct OperandsSignature {
                            bool StripImmCodes = false) const {
-    for (unsigned i = 0, e = Operands.size(); i != e; ++i)
-      Operands[i].printManglingSuffix(OS, ImmPredicates, StripImmCodes);
+    for (const OpKind Opnd : Operands)
```
Here and everywhere: we don't usually use `const` for non-pointer non-reference types.
Personally, using `const` like this makes me wonder whether this is indented or you've forgotten to add `&`. IOW it is an unnecessary cognitive load.
Since this is not standardized either, my argument would be "omit `const` for consistency with existing code".

jurahul replied:
Done

s-barannikov reviewed (commented):
Please do *never* amend commits after reviewers left comments. This makes it much more difficult to do incremental review, or (when combined with rebase) practically impossible.
Constant rebases aren't needed either. If you really need to update the branch (to resolve merge conflicts is the only valid reason), press the "Updata branch" button in GitHub UI.

jurahul commented:
> Please do _never_ amend commits after reviewers left comments. This makes it much more difficult to do incremental review, or (when combined with rebase) practically impossible. Constant rebases aren't needed either. If you really need to update the branch (to resolve merge conflicts is the only valid reason), press the "Updata branch" button in GitHub UI.

Agreed. I don't think I did it for this PR, unless I am missing something. But I might have done for some earlier PRs as a part of addressing review comments. In that case I do tend to rebase on top of latest changes and then apply the additional delta to address review comments, and I can see how that can impede incremental reviews. I'll be more mindful going forward and refrain from doing that.

s-barannikov commented:
> I don't think I did it for this PR, unless I am missing something.

No, but I thought I should warn you before your do :)

> But I might have done for some earlier PRs as a part of addressing review comments. In that case I do tend to rebase on top of latest changes and then apply the additional delta to address review comments, and I can see how that can impede incremental reviews. I'll be more mindful going forward and refrain from doing that.

Thanks a lot!

jurahul committed:
Review feedback

jurahul commented:
I pushed the delta into a new commit and did not rebase, hopefully that helps review just the new delta

s-barannikov commented on llvm/utils/TableGen/FastISelEmitter.cpp:
```diff
@@ -151,48 +148,43 @@ struct OperandsSignature {
-      else
-        Result.Operands.push_back(OpKind::getImm(0));
+    Result.Operands.resize(Operands.size());
+    llvm::transform(Operands, Result.Operands.begin(), [](OpKind Kind) {
```
I think this should be:
```suggestion
    llvm::transform(Operands, std::back_inserter(Result.Operands), [](OpKind Kind) {
```
Otherwise it will overwrite `*begin()` every time (?)

s-barannikov replied:
No, I'm wrong. It does increment the result iterator.

s-barannikov commented on llvm/utils/TableGen/FastISelEmitter.cpp:
```diff
@@ -151,48 +148,43 @@ struct OperandsSignature {
-        Result.Operands.push_back(Operands[i]);
-      else
-        Result.Operands.push_back(OpKind::getImm(0));
+    Result.Operands.resize(Operands.size());
```
With the below suggestion:
```suggestion
    Result.Operands.reserve(Operands.size());
```
